### ISSUE TITLE
Access to Emscripten's file system directly using Module.FS

### ISSUE THREAD

**Issue by madhur-tandon** [pyodide/pyodide#3274]
## Feature

Access the Emscripten Filesystem directly instead of waiting for `pyodide` to initialise and be ready first.

### Motivation

The Emscripten File system is provided by using the flag `-s 'EXPORTED_RUNTIME_METHODS=["FS"]'` when compiling C programs to WASM using `emcc`. This provides a direct access to the file system using `Module.FS`. 

In case of `pyodide` however, because of the way it is compiled, the file system is accessible through `pyodide.FS` (which is just a re-export of `Module.FS`) or perhaps through `pyodide._module.FS` as mentioned here:

https://pyodide.org/en/latest/usage/file-system.html#dealing-with-the-file-system

While there is no difference apart from the way to access it, the side effect is of waiting for `pyodide` to be ready first.

### Pitch

It would be nice to have access to `Module.FS` directly, which has the advantage of not waiting on the initialisation / startup time. One use-case could be fetching and setting up any additional non-pyodide specific files in parallel even before the runtime is ready.

Further, envisioning that more runtimes / interpreters (such as Micropython) will / are adopting WebAssembly, this provides a consistent way of accessing the File System across all such runtimes / interpreters.

### Alternatives

This is just meant to initiate a discussion and perhaps there are better ways to think about this. And there might be some caveats in this approach which are being overlooked.

### Additional context

Tagging some folks from PyScript here: @fpliger @antocuni @ntoll and from Pyodide as well: @hoodmane @rth

**Comment by antocuni** [comment]
For more pyscript-specific context which might be helpful to understand why it would be useful for us: currently pyscript has a way to fetch files from external URLs and save the to the virtual FS. A very simple example is this:
```html
<py-config>
[[fetch]]
files = ["foo.py", "bar/__init__.py", "bar/baz.py"]
</py-config>
```

Currently we need to wait for `loadPyodide` to finish before we can start writing them to the FS.
I think that with enough workarounds we could probably find a way to "fetch-early/write-late", but having access to the FS in advance sounds cleaner (and might be generally useful anyway).

**Comment by rth** [comment]
Thanks for the proposal! As far as I remember the initial motivation was to avoid creating global variables, and that having everything in the pyodide object is better from the namespacing point of view. In particular, defining global variables might be a problem when there are multiple Pyodide instances loaded, but Hood would probably have more thoughts about this.

> we would like to do (2) and (3) in parallel

OK, but you would still not have access to the FS until at least the beginning of bootstrapping in `loadPyodide` that initializes the Emscripten runtime with the FS, as far as I understand. So even if this is implemented you won't be able to run them fully in parallel, without some way to wait until the FS is available anyway?

The other thing is that as you said, you still can fetch the files async in parallel (which should be the main bottleneck), then write them when the pyodide objects is created. So in the end you would potentially be only saving the time to write those files to the FS which should be almost negligible (unless they are big files)?

**Comment by ntoll** [comment]
> The other thing is that as you said, you still can fetch the files async in parallel (which should be the main bottleneck), then write them when the pyodide objects is created. So in the end you would potentially be only saving the time to write those files to the FS which should be almost negligible (unless they are big files)?

This is pretty much what MicroPyScript does with its pluggable runtimes (including Pyodide). See: https://github.com/pyscript/MicroPyScript/blob/main/pyscript.js#L649

**Comment by hoodmane** [comment]
There are some use cases for being able to make early file operations including:
1. Making a virtual environment (I added a special private argument to `loadPyodide` to allow this in the command line runner)
2. Modifying the stdlib

Another thing that would be useful would be if you could pass a list of package names for Pyodide to begin downloading as it's setting up the runtime. This would improve loading efficiency over the only current option which is completely setting up the Pyodide runtime and only then beginning to download packages.

**Comment by antocuni** [comment]
re @rth 

> the initial motivation was to avoid creating global variables, and that having everything in the pyodide object is better from the namespacing point of view.

yes, +1 for avoiding cluttering the global namespace. Some random (and probably not well thought) ideas:

1. attach an `FS` attribute to `loadPyodide`: so you could do `loadPyodide.FS`, and `loadPyodide().FS === loadPyodide.FS`. I'm not sure I like this to be honest

2. add a compile-time option to choose whether to export the `FS` globally, and which name to use. I think that in this case pyscript could just use its own custom build of pyodide

3. change the API: instead of offering a single `loadPyodide` function, we could offer a global object with many attributes; e.g. `Pyodide.load()`, `Pyodide.FS`. For few releases, `loadPyodide` would still exist and be a synonym for `Pyodide.load`. The exact naming of the global object can be discussed of course: `pyodide`, `Pyodide`, `PYODIDE`, `pyodideAPI`?

3a. if we go in this direction, maybe we could consider a different name than `load`, e.g. `Pyodide.create()`, `.launch`, etc. Personally, I think that `load` is a bit too overloaded and risk to cause confusion (e.g. there is already the `load` DOM event, but also `pyodide.loadPackage` which does something different).


> OK, but you would still not have access to the FS until at least the beginning of bootstrapping in `loadPyodide` that initializes the Emscripten runtime with the FS, as far as I understand. So even if this is implemented you won't be able to run them fully in parallel, without some way to wait until the FS is available anyway?

ah, I wasn't aware of that. I assume that the FS is initialized and available immediately (and I imagine that the FS initialization is so simple/quick that it doesn't really matter). Where is the relevant pyodide code which initializes the FS at startup?

---

re @hoodmane 

> Another thing that would be useful would be if you could pass a list of package names for Pyodide to begin downloading as it's setting up the runtime. This would improve loading efficiency over the only current option which is completely setting up the Pyodide runtime and only then beginning to download packages.

this would be useful but not a complete solution, because you need to fully resolve the environment to know the complete list of files which you need to download. For pyscript, I have a very rough idea of offering the ability of freezing the app, which would precompute the environment and somehow embed the complete list of URLs that you need to download. But again, we would need early access to the FS to do that.

**Comment by hoodmane** [comment]
Well we removed global variables so that you can use `loadPyodide` multiple times. Each instance of Pyodide loaded in this way gets its own file system. Most of the options you've suggested would be a bit strange in the case where you call `loadPyodide` more than once this.

> > pass a list of package names for Pyodide to begin downloading as it's setting up the runtime
> this would be useful but not a complete solution, because you need to fully resolve the environment to know the complete list of files which you need to download.

In that case we could accept a promise for the list of packages you want downloaded. Then if you have an async function that resolves the environment you can pass that. If you need to use Python code to decide what you need to download, then you have to wait for pretty much everything be be loaded anyways.

Another more flexible option would be to accept a `preRun` callback:
```js
	FSinitCallback : async function(FS) {
		const data = await get_data()
		FS.writeFile("/some/path", data);
	}
```
the implementation of this would look like:
```js
let extraFSInitPromise = Promise.resolve();
Module.preRun.push(
	() => {
		extraFSInitPromise = config.FSinitCallback(Module.FS);
	}
);
```
and then we would need to `await extraFSInitPromise` before `loadPyodide` returns. We would have to make some minor rearrangement if we wanted to ensure that `FSinitCallback` resolves prior to initializing Python which would be a useful option for e.g., setting up a virtual environment.

I think it would be good to add both a list of packages to load on startup and a callback to put extra stuff into the file system before Python initialization. Both of these won't be able to execute Python code.

> I assume that the FS is initialized and available immediately (and I imagine that the FS initialization is so simple/quick that it doesn't really matter). Where is the relevant pyodide code which initializes the FS at startup?

The file system is not available immediately. The initialization happens inside of `_createPyodideModule`. Initializing it consists of three steps:
1. bringing in a large chunk of code
2. calling a small amount of initialization code mostly to setup stdin, stdout and stderr (function is called `createStandardStreams()`).
3. loading `pyodide.asm.data` and LZ4 decompressing it to populate the file system with the Python standard library in `/lib/python3.10`

`_createPyodideModule` calls into `pyodide.asm.js` which contains all the code produced by Emscripten (and most of our JavaScript code as well). The code stretches from line 12228 to line 14051 here:
https://cdn.jsdelivr.net/pyodide/v0.22.0a2/debug/pyodide.asm.js 
Most of this code comes from the following Emscripten source file:
https://github.com/emscripten-core/emscripten/blob/main/src/library_fs.js

**Comment by ryanking13** [comment]
Perhaps we can add `module` argument that can be passed to `loadPyodide` so that users who know the Emscripten initialization process can add any hooks they want. For example, users can add [`preInit`](https://emscripten.org/docs/api_reference/module.html#Module.preInit) hook which can be used to initialize the file system.

Or another fetch-early/write-later idea would be mounting directories after initialization: `pyodide.FS.mount(MEMFS, {...}, "<path>")`

**Comment by rth** [comment]
+1 to make it possible to use Emscripten's `preRun`, `preInit` hooks. As far as I understand that would address your use case, right?

> Perhaps we can add module argument that can be passed to loadPyodide

Maybe a `moduleInitCallback` that takes the `Module` argument and could then set any of the emscripten hooks? For instance it could be called [here](https://github.com/pyodide/pyodide/blob/a00f871eef0a5e7f663831f30b3d42175172a123/src/js/pyodide.ts#L312). That would be more general than FSinitCallback.  And nested callback definition is not the best thing, but it would be very flexible and backward compatible.

**Comment by antocuni** [comment]
> Well we removed global variables so that you can use loadPyodide multiple times. Each instance of Pyodide loaded in this way gets its own file system

oh, I wasn't aware of that. I thought/assumed that the FS is a singleton ans it's shared between the various emscripten "processes". I agree, if the FS is per-pyodide, then my suggestions are non-sense, sorry for that :)

---

> +1 to make it possible to use Emscripten's `preRun`, `preInit` hooks. As far as I understand that would address your use case, right?

yes, I think that given the other constraints this is probably the best way to solve our use case (although it makes the code on pyscript side a bit more complicated than what I would have liked, but too bad :)).

**Comment by hoodmane** [comment]
It would be interesting to see if we could share one `FS` instance between multiple Pyodide instances. In a lot of ways that makes more sense than the current system. Either sharing the FS or allowing users to shut down and restart the interpreter might help all the people asking to reinstall / reload modules.
